activerecord/lib/active_record/connection_adapters/abstract_mysql_adapter.rb

```diff
@@ -49,7 +49,7 @@ class AbstractMysqlAdapter < AbstractAdapter
       class StatementPool < ConnectionAdapters::StatementPool # :nodoc:
         private
           def dealloc(stmt)
-            stmt.close
+            stmt.close if stmt
```

Comment:
I don't get how `stmt` could be `nil` here.

Reply:
I get:

```
ActiveRecord::StatementInvalid: NoMethodError: undefined method `close' for nil
    lib/active_record/connection_adapters/abstract_mysql_adapter.rb:52:in `dealloc'
```

<details>
<summary>Full stacktraces:</summary>

```

Failure:
ActiveRecord::AdapterConnectionTest#test_queries_containing_SQL_fragments_are_not_retried [test/cases/adapter_test.rb:728]:
[ActiveRecord::ConnectionFailed] exception expected, not
Class: <ActiveRecord::StatementInvalid>
Message: <"NoMethodError: undefined method `close' for nil">
---Backtrace---
/home/zzak/code/rails/activerecord/lib/active_record/connection_adapters/abstract_mysql_adapter.rb:52:in `dealloc'
/home/zzak/code/rails/activerecord/lib/active_record/connection_adapters/statement_pool.rb:53:in `delete'
/home/zzak/code/rails/activerecord/lib/active_record/connection_adapters/mysql2/database_statements.rb:84:in `rescue in perform_query'
/home/zzak/code/rails/activerecord/lib/active_record/connection_adapters/mysql2/database_statements.rb:64:in `perform_query'
/home/zzak/code/rails/activerecord/lib/active_record/connection_adapters/abstract/database_statements.rb:556:in `block (2 levels) in raw_execute'
/home/zzak/code/rails/activerecord/lib/active_record/connection_adapters/abstract_adapter.rb:1029:in `block in with_raw_connection'
/home/zzak/code/rails/activesupport/lib/active_support/concurrency/null_lock.rb:9:in `synchronize'
/home/zzak/code/rails/activerecord/lib/active_record/connection_adapters/abstract_adapter.rb:998:in `with_raw_connection'
/home/zzak/code/rails/activerecord/lib/active_record/connection_adapters/abstract/database_statements.rb:555:in `block in raw_execute'
/home/zzak/code/rails/activesupport/lib/active_support/notifications/instrumenter.rb:58:in `instrument'
/home/zzak/code/rails/activerecord/lib/active_record/connection_adapters/abstract_adapter.rb:1149:in `log'
/home/zzak/code/rails/activerecord/lib/active_record/connection_adapters/abstract/database_statements.rb:554:in `raw_execute'
/home/zzak/code/rails/activerecord/lib/active_record/connection_adapters/abstract/database_statements.rb:591:in `internal_execute'
/home/zzak/code/rails/activerecord/lib/active_record/connection_adapters/abstract/database_statements.rb:547:in `internal_exec_query'
/home/zzak/code/rails/activerecord/lib/active_record/connection_adapters/abstract/database_statements.rb:693:in `select'
/home/zzak/code/rails/activerecord/lib/active_record/connection_adapters/abstract/database_statements.rb:73:in `select_all'
/home/zzak/code/rails/activerecord/lib/active_record/connection_adapters/abstract/query_cache.rb:251:in `select_all'
/home/zzak/code/rails/activerecord/lib/active_record/connection_adapters/mysql2/database_statements.rb:12:in `select_all'
/home/zzak/code/rails/activerecord/lib/active_record/querying.rb:68:in `_query_by_sql'
/home/zzak/code/rails/activerecord/lib/active_record/relation.rb:1449:in `block (2 levels) in exec_main_query'
/home/zzak/code/rails/activerecord/lib/active_record/connection_adapters/abstract/connection_pool.rb:412:in `with_connection'
/home/zzak/code/rails/activerecord/lib/active_record/connection_handling.rb:310:in `with_connection'
/home/zzak/code/rails/activerecord/lib/active_record/relation.rb:1448:in `block in exec_main_query'
/home/zzak/code/rails/activerecord/lib/active_record/relation.rb:1470:in `skip_query_cache_if_necessary'
/home/zzak/code/rails/activerecord/lib/active_record/relation.rb:1432:in `exec_main_query'
/home/zzak/code/rails/activerecord/lib/active_record/relation.rb:1410:in `block in exec_queries'
/home/zzak/code/rails/activerecord/lib/active_record/relation.rb:1470:in `skip_query_cache_if_necessary'
/home/zzak/code/rails/activerecord/lib/active_record/relation.rb:1404:in `exec_queries'
/home/zzak/code/rails/activerecord/lib/active_record/relation.rb:1181:in `load'
/home/zzak/code/rails/activerecord/lib/active_record/relation.rb:343:in `records'
/home/zzak/code/rails/activerecord/lib/active_record/relation/finder_methods.rb:586:in `find_take'
/home/zzak/code/rails/activerecord/lib/active_record/relation/finder_methods.rb:129:in `take'
/home/zzak/code/rails/activerecord/lib/active_record/relation/finder_methods.rb:112:in `find_by'
/home/zzak/code/rails/activerecord/lib/active_record/querying.rb:24:in `find_by'
/home/zzak/code/rails/activerecord/lib/active_record/core.rb:276:in `find_by'
/home/zzak/code/rails/activerecord/lib/active_record/encryption/extended_deterministic_queries.rb:129:in `find_by'
/home/zzak/code/rails/activerecord/test/cases/adapter_test.rb:728:in `block (2 levels) in <class:AdapterConnectionTest>'
---------------


bin/rails test /home/zzak/code/rails/activerecord/test/cases/adapter_test.rb:713

.......F

Failure:
ActiveRecord::AdapterConnectionTest#test_queries_containing_SQL_functions_are_not_retried [test/cases/adapter_test.rb:740]:
[ActiveRecord::ConnectionFailed] exception expected, not
Class: <ActiveRecord::StatementInvalid>
Message: <"NoMethodError: undefined method `close' for nil">
---Backtrace---
/home/zzak/code/rails/activerecord/lib/active_record/connection_adapters/abstract_mysql_adapter.rb:52:in `dealloc'
/home/zzak/code/rails/activerecord/lib/active_record/connection_adapters/statement_pool.rb:53:in `delete'
/home/zzak/code/rails/activerecord/lib/active_record/connection_adapters/mysql2/database_statements.rb:84:in `rescue in perform_query'
/home/zzak/code/rails/activerecord/lib/active_record/connection_adapters/mysql2/database_statements.rb:64:in `perform_query'
/home/zzak/code/rails/activerecord/lib/active_record/connection_adapters/abstract/database_statements.rb:556:in `block (2 levels) in raw_execute'
/home/zzak/code/rails/activerecord/lib/active_record/connection_adapters/abstract_adapter.rb:1029:in `block in with_raw_connection'
/home/zzak/code/rails/activesupport/lib/active_support/concurrency/null_lock.rb:9:in `synchronize'
/home/zzak/code/rails/activerecord/lib/active_record/connection_adapters/abstract_adapter.rb:998:in `with_raw_connection'
/home/zzak/code/rails/activerecord/lib/active_record/connection_adapters/abstract/database_statements.rb:555:in `block in raw_execute'
/home/zzak/code/rails/activesupport/lib/active_support/notifications/instrumenter.rb:58:in `instrument'
/home/zzak/code/rails/activerecord/lib/active_record/connection_adapters/abstract_adapter.rb:1149:in `log'
/home/zzak/code/rails/activerecord/lib/active_record/connection_adapters/abstract/database_statements.rb:554:in `raw_execute'
/home/zzak/code/rails/activerecord/lib/active_record/connection_adapters/abstract/database_statements.rb:591:in `internal_execute'
/home/zzak/code/rails/activerecord/lib/active_record/connection_adapters/abstract/database_statements.rb:547:in `internal_exec_query'
/home/zzak/code/rails/activerecord/lib/active_record/connection_adapters/abstract/database_statements.rb:693:in `select'
/home/zzak/code/rails/activerecord/lib/active_record/connection_adapters/abstract/database_statements.rb:73:in `select_all'
/home/zzak/code/rails/activerecord/lib/active_record/connection_adapters/abstract/query_cache.rb:251:in `select_all'
/home/zzak/code/rails/activerecord/lib/active_record/connection_adapters/mysql2/database_statements.rb:12:in `select_all'
/home/zzak/code/rails/activerecord/lib/active_record/querying.rb:68:in `_query_by_sql'
/home/zzak/code/rails/activerecord/lib/active_record/relation.rb:1449:in `block (2 levels) in exec_main_query'
/home/zzak/code/rails/activerecord/lib/active_record/connection_adapters/abstract/connection_pool.rb:412:in `with_connection'
/home/zzak/code/rails/activerecord/lib/active_record/connection_handling.rb:310:in `with_connection'
/home/zzak/code/rails/activerecord/lib/active_record/relation.rb:1448:in `block in exec_main_query'
/home/zzak/code/rails/activerecord/lib/active_record/relation.rb:1470:in `skip_query_cache_if_necessary'
/home/zzak/code/rails/activerecord/lib/active_record/relation.rb:1432:in `exec_main_query'
/home/zzak/code/rails/activerecord/lib/active_record/relation.rb:1410:in `block in exec_queries'
/home/zzak/code/rails/activerecord/lib/active_record/relation.rb:1470:in `skip_query_cache_if_necessary'
/home/zzak/code/rails/activerecord/lib/active_record/relation.rb:1404:in `exec_queries'
/home/zzak/code/rails/activerecord/lib/active_record/relation.rb:1181:in `load'
/home/zzak/code/rails/activerecord/lib/active_record/relation.rb:343:in `records'
/home/zzak/code/rails/activerecord/lib/active_record/relation.rb:338:in `to_ary'
/home/zzak/code/rails/activerecord/lib/active_record/relation/finder_methods.rb:615:in `find_nth_with_limit'
/home/zzak/code/rails/activerecord/lib/active_record/relation/finder_methods.rb:600:in `find_nth'
/home/zzak/code/rails/activerecord/lib/active_record/relation/finder_methods.rb:177:in `first'
/home/zzak/code/rails/activerecord/test/cases/adapter_test.rb:741:in `block (2 levels) in <class:AdapterConnectionTest>'
---------------


bin/rails test /home/zzak/code/rails/activerecord/test/cases/adapter_test.rb:732

.........E

Error:
ActiveRecord::AdapterConnectionTest#test_#find_and_#find_by_queries_with_known_attributes_are_retried_and_result_in_a_reconnect:
ActiveRecord::StatementInvalid: NoMethodError: undefined method `close' for nil
    lib/active_record/connection_adapters/abstract_mysql_adapter.rb:52:in `dealloc'
    lib/active_record/connection_adapters/statement_pool.rb:53:in `delete'
    lib/active_record/connection_adapters/mysql2/database_statements.rb:84:in `rescue in perform_query'
    lib/active_record/connection_adapters/mysql2/database_statements.rb:64:in `perform_query'
    lib/active_record/connection_adapters/abstract/database_statements.rb:556:in `block (2 levels) in raw_execute'
    lib/active_record/connection_adapters/abstract_adapter.rb:1029:in `block in with_raw_connection'
    /home/zzak/code/rails/activesupport/lib/active_support/concurrency/null_lock.rb:9:in `synchronize'
    lib/active_record/connection_adapters/abstract_adapter.rb:998:in `with_raw_connection'
    lib/active_record/connection_adapters/abstract/database_statements.rb:555:in `block in raw_execute'
    /home/zzak/code/rails/activesupport/lib/active_support/notifications/instrumenter.rb:58:in `instrument'
    lib/active_record/connection_adapters/abstract_adapter.rb:1149:in `log'
    lib/active_record/connection_adapters/abstract/database_statements.rb:554:in `raw_execute'
    lib/active_record/connection_adapters/abstract/database_statements.rb:591:in `internal_execute'
    lib/active_record/connection_adapters/abstract/database_statements.rb:547:in `internal_exec_query'
    lib/active_record/connection_adapters/abstract/database_statements.rb:693:in `select'
    lib/active_record/connection_adapters/abstract/database_statements.rb:73:in `select_all'
    lib/active_record/connection_adapters/abstract/query_cache.rb:251:in `select_all'
    lib/active_record/connection_adapters/mysql2/database_statements.rb:12:in `select_all'
    lib/active_record/querying.rb:68:in `_query_by_sql'
    lib/active_record/querying.rb:53:in `block in find_by_sql'
    lib/active_record/connection_adapters/abstract/connection_pool.rb:412:in `with_connection'
    lib/active_record/connection_handling.rb:310:in `with_connection'
    lib/active_record/querying.rb:52:in `find_by_sql'
    lib/active_record/statement_cache.rb:152:in `execute'
    lib/active_record/core.rb:443:in `block in cached_find_by'
    lib/active_record/connection_adapters/abstract/connection_pool.rb:412:in `with_connection'
    lib/active_record/connection_handling.rb:310:in `with_connection'
    lib/active_record/core.rb:431:in `cached_find_by'
    lib/active_record/core.rb:268:in `find'
    test/cases/adapter_test.rb:704:in `block in <class:AdapterConnectionTest>'


bin/rails test /home/zzak/code/rails/activerecord/test/cases/adapter_test.rb:699

```

Reply:
I see, then the bug is in statement_pool, and should be handled there:

```ruby
      def delete(key)
        dealloc cache[key]
        cache.delete(key)
      end
```

We wrongly assume here that `key` exist.

Reply:
https://github.com/rails/rails/pull/53870